lib/tests/streamlit/runtime/state/query_params_test.py

```diff
@@ -0,0 +1,138 @@
+            self._query_params.nonexistent
+
+    def test__getitem__returns_last_element_of_list(self):
+        assert self._query_params["two"] == "y"
+
+    def test__getitem__converts_int_value_to_string(self):
+        self._query_params["baz"] = 1
+        assert self._query_params["baz"] == "1"
+
+    def test__getitem__converts_float_value_to_string(self):
+        self._query_params["corge"] = 1.23
+        assert self._query_params["corge"] == "1.23"
+
+    def test__getitem__retrieves_existing_key(self):
+        assert self._query_params["foo"] == "bar"
+
+    def test_get_method_retrieves_existing_key(self):
+        assert self._query_params.get("foo") == "bar"
+
+    def test_get_method_returns_default_for_nonexistent_key(self):
+        return_val = self._query_params.get("dog", default="bark")
+        assert return_val == "bark"
+
+    def test__getitem__retrieves_existing_key(self):
+        assert self._query_params.foo == "bar"
+
+    def test__setitem__adds_new_query_param(self):
+        assert "test" not in self._query_params
+        self._query_params["test"] = "test"
+        assert self._query_params.get("test") == "test"
+        message = self.get_message_from_queue(0)
+        assert "test=test" in message.page_info_changed.query_string
+
+    def test__setitem__handles_empty_string_value(self):
+        assert "test" not in self._query_params
+        self._query_params["test"] = ""
+        assert self._query_params["test"] == ""
+        message = self.get_message_from_queue(0)
+        assert "test=" in message.page_info_changed.query_string
+
+    def test__setattr__adds_new_query_param(self):
+        self._query_params.test = "test"
+        assert self._query_params.get("test") == "test"
+
+    def test__setattr__sets_old_query_param_key(self):
+        self._query_params.foo = "test"
+        assert self._query_params.get("foo") == "test"
+
+    def test__setattr__adds_list_value(self):
+        self._query_params.test = ["test", "test2"]
+        assert self._query_params["test"] == "test2"
+        message = self.get_message_from_queue(0)
+        assert "test=test&test=test2" in message.page_info_changed.query_string
+
+    def test_get_all_returns_empty_list_for_nonexistent_key(self):
+        assert self._query_params.get_all("nonexistent") == []
+
+    def test_get_all_retrieves_single_element_list(self):
+        assert self._query_params.get_all("foo") == ["bar"]
+
+    def test_get_all_retrieves_multiple_values_as_list(self):
+        assert self._query_params.get_all("two") == ["x", "y"]
+
+    def test_get_all_handles_mixed_type_values(self):
+        self._query_params["test"] = ["", "a", 1, 1.23]
+        assert self._query_params.get_all("test") == ["", "a", "1", "1.23"]
+
+    def test__contains__returns_true_for_present_key(self):
+        assert "foo" in self._query_params
+
+    def test__contains__returns_false_for_absent_key(self):
+        assert "baz" not in self._query_params
+
```

Comment:
Same content for these just being a test of `MutableMapping` assuming that we get `__contains__` implemented for free